packages/canary-web-components/src/utils/helpers.ts

```diff
@@ -664,15 +664,18 @@ export const addDataToPosition = (
 ): IcDataTableDataType => {
   const newData: IcDataTableDataType = {};
   const newIndexes = newKeys.map((key) => key.index);
-  let controlledIndex = 0; // When a new key is added to the data, need to increment the index to account for this new object value
 
-  Object.keys(dataObject).forEach((dataKey) => {
-    if (newIndexes.includes(controlledIndex)) {
```

Comment:
this didn't work if e.g. new indexes were 0,1 as 1 would never be called as controlledIndex would skip over it.

Also didn't work for end columns if there were two slotted columns as the object only iterated over the existing keys and it finished iterating before it got to the end column